Source/WebCore/loader/CrossOriginOpenerPolicy.cpp

```diff
@@ -193,20 +193,27 @@ CrossOriginOpenerPolicy obtainCrossOriginOpenerPolicy(const ResourceResponse& re
         return *coep;
     };
     auto parseCOOP = [&response, &ensureCOEP](HTTPHeaderName headerName, auto& value, auto& reportingEndpoint) {
-        auto coopParsingResult = parseStructuredFieldValue(response.httpHeaderField(headerName));
+        std::optional<std::pair<RFC8941::BareItem, RFC8941::Parameters>> coopParsingResult = RFC8941::parseItemStructuredFieldValue(response.httpHeaderField(headerName));
         if (!coopParsingResult)
             return;
 
-        if (coopParsingResult->first == "same-origin"_s) {
+        auto* policyString = std::get_if<RFC8941::Token>(&coopParsingResult->first);
+        if (!policyString)
+            return;
+
+        if (policyString->string() == "same-origin"_s) {
             auto& coep = ensureCOEP();
             if (coep.value == CrossOriginEmbedderPolicyValue::RequireCORP || (headerName == HTTPHeaderName::CrossOriginOpenerPolicyReportOnly && coep.reportOnlyValue == CrossOriginEmbedderPolicyValue::RequireCORP))
                 value = CrossOriginOpenerPolicyValue::SameOriginPlusCOEP;
             else
                 value = CrossOriginOpenerPolicyValue::SameOrigin;
-        } else if (coopParsingResult->first == "same-origin-allow-popups"_s)
+        } else if (policyString->string() == "same-origin-allow-popups"_s)
             value = CrossOriginOpenerPolicyValue::SameOriginAllowPopups;
 
-        reportingEndpoint = coopParsingResult->second.get<HashTranslatorASCIILiteral>("report-to"_s);
+        if (auto it = coopParsingResult->second.find<HashTranslatorASCIILiteral>("report-to"_s); it != coopParsingResult->second.end()) {
```

Comment:
This little bit of logic is complicated enough and similar enough to the one in CrossOriginEmbedderPolicy.cpp that it would good to see if it can be lifted into its own function somewhere, maybe by wrapping the RFC8941::Parameters hash map in a struct and adding some helpers like this?

Reply:
Ok, I'll look into it.